File: src/minkowski_diff.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <string>
#include "matplotlibcpp.h"
#define _USE_MATH_DEFINES
namespace plt = matplotlibcpp;

struct Point{
    float x;
    float y;
};

std::vector<Point> Minkowski_diff(std::vector<Point> a, std::vector<Point> b);
std::vector<std::vector<float>> Point2Vec(std::vector<Point> a);
float angle(Point a, Point b);
std::vector<Point> move(std::vector<Point> a, std::vector<Point> b);
std::vector<std::vector<Point>> rotatePoints(std::vector<Point> robot, int partitions);

void part_a();
void part_b();

int main(){
    part_a();
    part_b();
    return 0;
}

void part_b(){
    int partitions = 10;
    std::vector<Point> obstacle             = std::vector<Point>{Point{0,0},Point{1,2},Point{0,2}};
    std::vector<Point> robot                = std::vector<Point>{Point{2,2},Point{3,4},Point{2,4}}; 
    std::vector<std::vector<Point>> rrobot  = rotatePoints(robot, partitions); // Points to be rotated and number of partitions of 360 degrees.
    int count = 0;
    for(std::vector<Point> rot:rrobot){
        std::vector<float> tt;
        std::vector<Point> Cspace_obs           = Minkowski_diff(obstacle, rot);
        std::vector<std::vector<float>> ans1    = Point2Vec(Cspace_obs);
        for(int i = 0; i < ans1[0].size(); i++) tt.push_back((2*M_PI/partitions)*count);
        std::cout << tt.size() << std::endl;
        plt::plot3(ans1[0], ans1[1], tt); 
        count++;
    }
    plt::grid(true);
    plt::legend();
    plt::xlabel("X axes");
    plt::ylabel("Y axes");
    plt::set_zlabel("Theta rotation");
    plt::title("C Space obstacle");
	plt::show();
}

void part_a(){
    std::vector<Point> obstacle             = std::vector<Point>{Point{0,0},Point{1,2},Point{0,2}};
    std::vector<Point> robot                = std::vector<Point>{Point{2,2},Point{3,4},Point{2,4}}; 
    std::vector<Point> Cspace_obs           = Minkowski_diff(obstacle, robot);
    std::vector<std::vector<float>> ans1    = Point2Vec(Cspace_obs);
    std::vector<std::vector<float>> ans2    = Point2Vec(obstacle);
    std::vector<std::vector<float>> ans3    = Point2Vec(robot);
    
    plt::named_plot("CSpace Obstacle", ans1[0], ans1[1],"*b");   
    // plt::named_plot("Trajectory", ans2[0], ans2[1],"*-r");
    // plt::named_plot("Trajectory", ans3[0], ans3[1],"*-g");
    plt::grid(true);
    plt::legend();
    plt::xlabel("X axes");
    plt::ylabel("Y axes");
    plt::title("C Space obstacle");
	plt::show();
}

std::vector<Point> Minkowski_diff(std::vector<Point> a, std::vector<Point> b){
    b               = move(b, a);
    std::vector<Point> minus_b;

    for(Point p: b){
        minus_b.push_back(Point{-p.x, -p.y});
    }

    std::vector<Point> diff_vec;
    int n = a.size()-1; int m = b.size()-1;
    // int i = 0;
    // int j = 0;
    for(int i = 0; i < a.size(); i++){
        for(int j = 0; j < b.size(); j++){
            Point Temp;
            Temp.x = a[i].x + minus_b[j].x;
            Temp.y = a[i].y + minus_b[j].y;
            diff_vec.push_back(Temp);
        }
    }
    // while(i < n+1 && j < m+1){
    //         Point Temp; 
    //         Temp.x = a[i].x + minus_b[j].x;
    //         Temp.y = a[i].y + minus_b[j].y;
    //         std::cout << i << " " << j << std::endl; 
    //         diff_vec.push_back(Temp);
    //         if(angle(a[i+1],a[i]) < angle(b[j+1],b[j])){
    //             i++;
    //         }
    //         else if(angle(a[i+1],a[i]) > angle(b[j+1],b[j])){
    //             j++;
    //         }
    //         else{
    //             i++; j++;
    //         }
    // }

    return diff_vec;
}

std::vector<std::vector<float>> Point2Vec(std::vector<Point> a){
    std::vector<std::vector<float>> ans{std::vector<float>{},std::vector<float>{}};
    for(Point p:a){
        ans[0].push_back(p.x);
        ans[1].push_back(p.y);
    }
    ans[0].push_back(a[0].x);
    ans[1].push_back(a[0].y);

    return ans;
}

float angle(Point a, Point b){
    return std::atan2(b.y - a.y, b.x - a.x);
}

std::vector<Point> move(std::vector<Point> a, std::vector<Point> b){
    float dt = -0.000001;
    float error = 999999;
    while(std::abs(error) >= 0.000001){
        float angle = std::atan2(b[0].y - a[0].y, b[0].x - a[0].x);
        for(int i = 0; i< a.size(); i++){
            a[i].x -= dt*std::cos(angle);
            a[i].y -= dt*std::sin(angle);
        }
        error = std::sqrt(std::pow(a[0].x - b[0].x,2) + std::pow(a[0].y - b[0].y,2));
    }
    return a;
}   


std::vector<std::vector<Point>> rotatePoints(std::vector<Point> robot, int partitions){
    std::vector<std::vector<Point>> ans;
    for(int i = 0; i < partitions; i++){
        std::vector<Point> temp;
        for(Point P:robot){
            Point tempP;
            tempP.x = P.x*std::cos(2*M_PI/partitions * i) - P.y*std::sin(2*M_PI/partitions * i);
            tempP.y = P.x*std::sin(2*M_PI/partitions * i) + P.y*std::cos(2*M_PI/partitions * i);
            temp.push_back(tempP);
        }
        ans.push_back(temp);
    }
    return ans;
}

File: src/minkowski_diff.py
import numpy as np
import math
from scipy.spatial import ConvexHull
import matplotlib.pyplot as plt
from mpl_toolkits import mplot3d


class point:
    x = 0.0
    y = 0.0
    def __init__(self,x_coord,y_coord) -> None:
        self.x = float(x_coord)
        self.y = float(y_coord)

def part_b():
    partitions = 12
    obstacle = [point(0.0,0.0),point(1.0,2.0),point(0.0,2.0)]
    robot = [point(0.0,0.0),point(1.0,2.0),point(0.0,2.0)]
    rrobot = rotatePoints(robot,partitions)
    fig = plt.figure()
    ax = plt.axes(projection='3d')
    tt = [i for i in np.arange(0,2*math.pi,2*math.pi/partitions)]
    count = 0
    for r in rrobot:
        Cspace_obs = minkowski_diff(obstacle,r)
        Cspace_x, Cspace_y = Point2list(Cspace_obs)
        ans = []
        for p in Cspace_obs:
            ans.append([p.x,p.y])
        ans.append([Cspace_obs[0].x, Cspace_obs[0].y])
        ans = np.array(ans)
        hull = ConvexHull(ans)
        for simplex in hull.simplices:
            ax.plot(ans[simplex, 0], ans[simplex, 1],tt[count], 'r-')
        z_val_pt = [tt[count] for _ in range(len(Cspace_x))]
        ax.plot3D(Cspace_x, Cspace_y,z_val_pt, '*b', lw=2)
        count+=1
    ax.set_xlabel('x')
    ax.set_ylabel('y')
    ax.set_zlabel('theta')
    plt.title("3D Cspace obstacles of two triangles one with rotation")
    plt.show()

def part_a():
    obstacle = [point(0.0,0.0),point(1.0,2.0),point(0.0,2.0)]
    robot = [point(2.0,2.0),point(3.0,4.0),point(2.0,4.0)]
    Cspace_obs = minkowski_diff(obstacle,robot)
    Cspace_x, Cspace_y = Point2list(Cspace_obs)
    ans = []
    for p in Cspace_obs:
        ans.append([p.x,p.y])
    ans.append([Cspace_obs[0].x, Cspace_obs[0].y])
    ans = np.array(ans)
    hull = ConvexHull(ans)
    for simplex in hull.simplices:
        plt.plot(ans[simplex, 0], ans[simplex, 1], 'r-')
    plt.plot(Cspace_x, Cspace_y,"*")
    plt.plot(ans[hull.vertices[0],0], ans[hull.vertices[0],1], 'ro')
    plt.title("2D Cspace obstacles of two triangles")
    plt.show()

def minkowski_diff(a,b):
    b = move(b,a)
    minus_b = []
    for p in b:
        minus_b.append(point(-p.x,-p.y))
    diff_vec = []
    for i in a:
        for j in minus_b:
            temp = point(0.0,0.0)
            diff_vec.append(point(i.x + j.x, i.y + j.y))
    return diff_vec


def move(a,b):
    dt = -0.0001
    error = 999999
    while(abs(error) >= 0.0001):
        angle = math.atan2(b[0].y - a[0].y, b[0].x - a[0].x)
        for i in range(len(a)):
            a[i].x -= dt*math.cos(angle)
            a[i].y -= dt*math.sin(angle)
        error = ((a[0].x - b[0].x)**2 + (a[0].y - b[0].y)**2)**0.5

    return a

def Point2list(a):
    ans_x, ans_y = [], []
    for p in a:
        ans_x.append(p.x)
        ans_y.append(p.y)
    
    return ans_x, ans_y


def rotatePoints(robot, partitions):
    ans = []
    for i in range(1,partitions):
        temp = []
        for P in robot:
            tempP = point(0.0,0.0)
            tempP.x = P.x*math.cos(2*math.pi/partitions * i) - P.y*math.sin(2*math.pi/partitions * i)
            tempP.y = P.x*math.sin(2*math.pi/partitions * i) + P.y*math.cos(2*math.pi/partitions * i)
            temp.append(tempP)
        ans.append(temp)
    return ans


part_a()
part_b()